### PR TITLE
Replace linker descriptor for AssemblyBuilder with DynamicDependency

### PR DESCRIPTION
**Pull request by marek-safar**
for better linking when AssemblyBuilder is not instantiated.

Contributes to #38692

@eerhardt

### CODE REVIEW

File: src/mono/netcore/System.Private.CoreLib/src/System/Reflection/Emit/AssemblyBuilder.Mono.cs [hunk 1]
```diff
@@ -219,6 +213,34 @@ public sealed class AssemblyBuilder : Assembly
...
+        [DynamicDependency("loaded_modules")]
+        [DynamicDependency("win32_resources")]
+        [DynamicDependency("permissions_minimum")]
+        [DynamicDependency("permissions_optional")]
+        [DynamicDependency("permissions_refused")]
+        [DynamicDependency("peKind")]
+        [DynamicDependency("machine")]
+        [DynamicDependency("corlib_internal")]
+        [DynamicDependency("type_forwarders")]
+        [DynamicDependency("pktoken")]
```
Comment (stephentoub) [inline src/mono/netcore/System.Private.CoreLib/src/System/Reflection/Emit/AssemblyBuilder.Mono.cs:243]: Nit: are you able to use nameof in any of these?

**Review by stephentoub — APPROVED** [review]

File: src/mono/netcore/System.Private.CoreLib/src/System/Reflection/Emit/AssemblyBuilder.Mono.cs [hunk 2]
```diff
@@ -202,13 +197,12 @@ public sealed class AssemblyBuilder : Assembly
         private object? permissions_minimum;
         private object? permissions_optional;
         private object? permissions_refused;
-        private PortableExecutableKinds peKind;
-        private ImageFileMachine machine;
+        private int peKind;
```
Comment (vargaz) [inline src/mono/netcore/System.Private.CoreLib/src/System/Reflection/Emit/AssemblyBuilder.Mono.cs:200]: Is this needed ?

**Review by vargaz — APPROVED** [review]

File: src/mono/netcore/System.Private.CoreLib/src/System/Reflection/Emit/AssemblyBuilder.Mono.cs (hunk 1 above)
Comment (vargaz) [reply src/mono/netcore/System.Private.CoreLib/src/System/Reflection/Emit/AssemblyBuilder.Mono.cs:243]: Could you a comment as to why the dependencies are on this method.

**Comment by eerhardt** [comment]
I tried using this change locally (with local fixes for #38693 and #38678). However, the ILLinker still isn't able to trim the fields from AssemblyBuilder. @marek-safar pointed out offline that it is because AssemblyBuilder has a `StructLayout` attribute.

https://github.com/dotnet/runtime/blob/a92f4f02dcbbd76cb9edf682424d001b9eb6d870/src/mono/netcore/System.Private.CoreLib/src/System/Reflection/Emit/AssemblyBuilder.Mono.cs#L170-L171

Even after changing that, there are other edges into Ref.Emit that are preserving most of the types (because of the other preserve=fields on other types):

```
TypeDef:System.Reflection.Emit.DynamicMethod size: 770
| Method:System.Delegate System.Delegate::CreateDelegate(System.Type,System.Object,System.Reflection.MethodInfo,System.Boolean,System.Boolean) [2 deps]
| Method:System.Delegate System.Delegate::CreateDelegate(System.Type,System.Reflection.MethodInfo,System.Boolean) [1 deps]
| Method:System.Delegate System.Delegate::CreateDelegate(System.Type,System.Reflection.MethodInfo) [3 deps]
| Method:System.Void Microsoft.AspNetCore.Components.Reflection.MemberAssignment/PropertySetter`2::.ctor(System.Reflection.MethodInfo,System.Boolean) [2 deps]
```

and 

```
 TypeDef:System.Reflection.Emit.ModuleBuilder size: 423
| Field:System.Reflection.Emit.ModuleBuilder System.Reflection.Emit.TypeBuilder::pmodule [1 deps]
| TypeDef:System.Reflection.Emit.TypeBuilder [72 deps]
| Method:System.Object System.Activator::CreateInstance(System.Type,System.Reflection.BindingFlags,System.Reflection.Binder,System.Object[],System.Globalization.CultureInfo,System.Object[]) [3 deps]
| Method:System.Object System.Activator::CreateInstance(System.Type,System.Object[]) [5 deps]
| Method:Microsoft.AspNetCore.Components.Reflection.IPropertySetter Microsoft.AspNetCore.Components.Reflection.MemberAssignment::CreatePropertySetter(System.Type,System.Reflection.PropertyInfo,System.Boolean) [3 deps]
```

**Comment by MichalStrehovsky** [comment]
I think this should be doable without any annotations.

`StructLayout.Sequential` keeps all fields on the type, for good reasons.

But linker is too eager to respect the layout. For reference types (and only reference types), the layout only becomes relevant once you allocate an instance. Linker should defer marking the fields until it sees the type as instantiated (i.e. constructor called).

Once linker does that, the annotation will be unnecessary and this will Just Work™.

**Comment by MichalStrehovsky** [comment]
> I think this should be doable without any annotations.

Linker change: mono/linker#1309

File: src/mono/netcore/System.Private.CoreLib/src/System/Reflection/Emit/AssemblyBuilder.Mono.cs (hunk 2 above)
Comment (marek-safar) [reply src/mono/netcore/System.Private.CoreLib/src/System/Reflection/Emit/AssemblyBuilder.Mono.cs:200]: the field is not used and it drops additional dependencies

File: src/mono/netcore/System.Private.CoreLib/src/System/Reflection/Emit/AssemblyBuilder.Mono.cs (hunk 1 above)
Comment (marek-safar) [reply src/mono/netcore/System.Private.CoreLib/src/System/Reflection/Emit/AssemblyBuilder.Mono.cs:243]: yeah, will switch to nameof

File: src/mono/netcore/System.Private.CoreLib/src/System/Reflection/Emit/AssemblyBuilder.Mono.cs [hunk 3]
```diff
@@ -224,7 +218,8 @@ public sealed class AssemblyBuilder : Assembly
         [MethodImplAttribute(MethodImplOptions.InternalCall)]
         private static extern void UpdateNativeCustomAttributes(AssemblyBuilder ab);
 
-        internal AssemblyBuilder(AssemblyName n, string? directory, AssemblyBuilderAccess access, bool corlib_internal)
+        [DynamicDependency(nameof(pktoken))] // Automatically keeps all previous fields too due to StructLayout
```
Comment (MichalStrehovsky) [inline src/mono/netcore/System.Private.CoreLib/src/System/Reflection/Emit/AssemblyBuilder.Mono.cs:221]: Why do we need DynamicDependency at all? The class is marked `StructLayout.Sequential` and this annotation is on a constructor - linker is going to keep all the fields anyway if it sees the constructor is needed.

Comment (marek-safar) [reply src/mono/netcore/System.Private.CoreLib/src/System/Reflection/Emit/AssemblyBuilder.Mono.cs:221]: Technically we don't right now but I wanted to make this dependency more explicit. We could also fix linker to keep only used fields for struct layout classes.

Comment (MichalStrehovsky) [reply src/mono/netcore/System.Private.CoreLib/src/System/Reflection/Emit/AssemblyBuilder.Mono.cs:221]: I'm not sure I understand how an explicit DynamicDependency on what appears to be an arbitrary field achieves that - I think we'll find that part quite confusing when we look at this code again 5 years from now.

Comment (stephentoub) [reply src/mono/netcore/System.Private.CoreLib/src/System/Reflection/Emit/AssemblyBuilder.Mono.cs:221]: If we want to be explicit, can we instead use the attribute that says to keep all fields?

**Pull request merged**